app/models/amazon_agent_manager.rb

```diff
@@ -0,0 +1,393 @@
+
+class AmazonAgentManager
+  include Comparable
+  include Vmdb::Logging
+  attr_accessor :ems, :deploying
+
+  MIQ_SSA   = "MIQ_SSA".freeze
+  RUBY_VERS = "2.3.3".freeze
+  AGENT_AMI = "centos-7.2-hvm*".freeze
+
+  def initialize(ems)
+    @ems = ems
+    @guid = MiqServer.my_server.guid
+
+    # List of active agent ids
+    @alive_agent_ids = []
+
+    # List of all agent ids, include those in power off state.
+    @agent_ids = []
+    @deploying = false
+  end
+
+  def ec2
+    @ec2 ||= ems.connect(:service => 'EC2')
+  end
+
+  def sqs
+    @sqs ||= ems.connect(:service => 'SQS')
+  end
+
+  def s3
+    @s3 ||= ems.connect(:service => 'S3')
+  end
+
+  def iam
+    @iam ||= ems.connect(:service => 'IAM')
+  end
+
+  def <=>(other)
+    @ems.guid <=> other.ems.guid
+  end
+
+  def agent_manager_settings
+    @agent_manager_settings ||= Settings.ems.ems_amazon.try(:agent_manager)
+  end
+
+  def region
+    @ems.provider_region
+  end
+
+  def log_level
+    ll = agent_manager_settings.try(:log_level) || AmazonSsaSupport::DEFAULT_LOG_LEVEL
+    ll.upcase
+  end
+
+  def heartbeat_prefix
+    agent_manager_settings.try(:heartbeat_prefix) || AmazonSsaSupport::DEFAULT_HEARTBEAT_PREFIX
+  end
+
+  def heartbeat_interval
+    agent_manager_settings.try(:heartbeat_interval) || AmazonSsaSupport::DEFAULT_HEARTBEAT_INTERVAL
+  end
+
+  def ssa_bucket
+    @ssa_bucket ||= (agent_manager_settings.try(:bucket_prefix) || AmazonSsaSupport::DEFAULT_BUCKET_PREFIX) + '-' + @ems.guid
+  end
+
+  def request_queue
+    @request_queue ||= (agent_manager_settings.try(:request_queue_prefix) || AmazonSsaSupport::DEFAULT_REQUEST_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_queue
+    @reply_queue ||= (agent_manager_settings.try(:reply_queue_prefix) || AmazonSsaSupport::DEFAULT_REPLY_QUEUE) + '-' + @ems.guid
+  end
+
+  def reply_prefix
+    agent_manager_settings.try(:reply_prefix) || AmazonSsaSupport::DEFAULT_REPLY_PREFIX
+  end
+
+  def log_prefix
+    agent_manager_settings.try(:log_prefix) || AmazonSsaSupport::DEFAULT_LOG_PREFIX
+  end
+
+  def agent_ami
+    agent_manager_settings.try(:agent_ami) || AGENT_AMI
+  end
+
+  def ruby_version
+    agent_manager_settings.try(:ruby_version) || RUBY_VERS
+  end
+
+  def userdata_script
+    agent_manager_settings.try(:userdata_script)
+  end
+
+  def profile_name
+    MIQ_SSA
+  end
+
+  def alive_agent_ids(interval = 180)
+    @alive_agent_ids = agent_ids.select { |id| agent_alive?(id, interval) }
+  end
+
+  def agent_ids
+    # reset to empty
+    @agent_ids = []
+
+    bucket = s3.bucket(ssa_bucket)
+    return @agent_ids unless bucket.exists?
+
+    bucket.objects(prefix: heartbeat_prefix).each do |obj|
+      id = obj.key.split('/')[2]
+      @agent_ids << id if ec2.instance(id).exists?
+    end
+
+    @agent_ids
+  end
+
+  def request_queue_empty?
+    messages_in_queue(request_queue).zero?
+  end
+
+  def reply_queue_empty?
+    messages_in_queue(reply_queue).zero?
+  end
+
+  def messages_in_queue(q_name)
```

Comment:
This looks like a private method.  In addition, it's only purpose seems to see if the queue is empty or not, so you can DRY up the .zero? as well, and name this method `message_queue_empty?`